File: backend/services/ai_service.py
"""
AI investigation service for Risk Sentinel.

In demo mode, produces structured responses based on risk engine output.
If an external AI API is configured (AI_API_KEY + AI_API_URL), it can be used.
The AI never invents transaction data — it only reasons over supplied signals.
All output uses cautious language: 'possible', 'potential', 'likely'.
"""
import os
from typing import Any


def _get_fallback_response(analysis: dict, txn: dict) -> dict:
    """Rule-based fallback when AI API is unavailable."""
    risk_score = analysis.get("risk_score", 0)
    risk_level = analysis.get("risk_level", "LOW")
    threat_type = analysis.get("threat_type", "Normal Transaction")
    recommended_action = analysis.get("recommended_action", "APPROVE")
    attack_story = analysis.get("attack_story", "")
    triggered = [f for f in analysis.get("risk_factors", []) if f.get("triggered")]
    txn_id = txn.get("id", "unknown")
    amount = txn.get("amount", 0)
    currency = txn.get("currency", "₹")
    location = txn.get("location", "Unknown")

    return {
        "risk_explanation": (
            f"Transaction {txn_id} for {currency}{int(amount):,} from {location} "
            f"was flagged with a risk score of {risk_score}/100 ({risk_level}). "
            f"{len(triggered)} risk signal(s) were triggered: "
            f"{', '.join(s['label'] for s in triggered) if triggered else 'none'}. "
            f"The most significant contributing factor is "
            f"{max(triggered, key=lambda x: x['deviation'])['label'] if triggered else 'behavioral deviation'}."
        ),
        "attack_story": attack_story,
        "threat_type": threat_type,
        "confidence": 0.87,
        "recommended_action": recommended_action,
        "investigation_summary": (
            f"Transaction {txn_id} | {currency}{int(amount):,} | {location} | "
            f"Score: {risk_score}/100 | Level: {risk_level} | "
            f"Threat: {threat_type} | Action: {recommended_action}"
        ),
        "demo": True,
    }


def get_ai_response(analysis: dict, txn: dict, question: str = "") -> dict:
    """
    Get AI investigation response.

    If AI_API_KEY and AI_API_URL are configured, calls external AI.
    Otherwise returns a structured demo/fallback response.
    """
    ai_api_key = os.environ.get("AI_API_KEY")
    ai_api_url = os.environ.get("AI_API_URL")

    if ai_api_key and ai_api_url:
        try:
            return _call_external_ai(ai_api_url, ai_api_key, analysis, txn, question)
        except Exception:
            return _get_fallback_response(analysis, txn)

    # Demo mode — structured responses based on question
    return _get_demo_response(analysis, txn, question)


def _get_demo_response(analysis: dict, txn: dict, question: str) -> dict:
    """Generate demo responses tailored to common questions."""
    risk_score = analysis.get("risk_score", 0)
    risk_level = analysis.get("risk_level", "LOW")
    threat_type = analysis.get("threat_type", "Normal Transaction")
    recommended_action = analysis.get("recommended_action", "APPROVE")
    attack_story = analysis.get("attack_story", "")
    triggered = [f for f in analysis.get("risk_factors", []) if f.get("triggered")]
    txn_id = txn.get("id", "unknown")
    amount = txn.get("amount", 0)
    currency = txn.get("currency", "₹")
    location = txn.get("location", "Unknown")

    q_lower = question.lower()

    if "flagged" in q_lower or "why" in q_lower:
        return {
            "risk_explanation": (
                f"This transaction was flagged because {', '.join(s['label'].lower() for s in triggered)} were detected. "
                f"The composite risk score is {risk_score}/100, classified as {risk_level}. "
                f"The most significant contributing signal is "
                f"{max(triggered, key=lambda x: x['deviation'])['label'] if triggered else 'behavioral deviation'}."
            ),
            "attack_story": attack_story,
            "threat_type": threat_type,
            "confidence": 0.87,
            "recommended_action": recommended_action,
            "investigation_summary": f"Transaction {txn_id} for {currency}{int(amount):,} from {location} triggered {len(triggered)} risk signals. Composite score: {risk_score}/100.",
            "demo": True,
        }

    if "strongest" in q_lower or "signal" in q_lower:
        top3 = sorted(triggered, key=lambda x: x["deviation"], reverse=True)[:3]
        return {
            "risk_explanation": (
                f"The strongest risk signals, ranked by deviation severity, are: "
                f"{', '.join(f'{i+1}. {s[\"label\"]} ({s[\"deviation\"]}% deviation)' for i, s in enumerate(top3))}. "
                f"These signals collectively contribute {sum(s['weight'] for s in top3)} points to the weighted risk score."
            ),
            "attack_story": attack_story,
            "threat_type": threat_type,
            "confidence": 0.91,
            "recommended_action": recommended_action,
            "investigation_summary": f"Top signals: {', '.join(s['label'] for s in top3)}.",
            "demo": True,
        }

    if "summar" in q_lower:
        return {
            "risk_explanation": (
                f"Investigation summary for transaction {txn_id}: A {currency}{int(amount):,} transaction from {location}. "
                f"Risk score: {risk_score}/100 ({risk_level}), {len(triggered)} triggered signals. "
                f"AI classifies this as potential {threat_type} and recommends {recommended_action}."
            ),
            "attack_story": attack_story,
            "threat_type": threat_type,
            "confidence": 0.89,
            "recommended_action": recommended_action,
            "investigation_summary": f"Transaction {txn_id} | {currency}{int(amount):,} | {location} | Score: {risk_score}/100 | {threat_type} | {recommended_action}",
            "demo": True,
        }

    if "attack" in q_lower or "pattern" in q_lower:
        return {
            "risk_explanation": (
                f"Based on the observed signal pattern — {', '.join(s['label'] for s in triggered)} — "
                f"the AI agent identifies a possible {threat_type}. "
                f"This pattern is consistent with known attack methodologies. "
                f"This is a probabilistic assessment, not a definitive determination."
            ),
            "attack_story": attack_story,
            "threat_type": threat_type,
            "confidence": 0.84,
            "recommended_action": recommended_action,
            "investigation_summary": f"Possible attack pattern: {threat_type}. Confidence: 84%.",
            "demo": True,
        }

    if "action" in q_lower or "recommend" in q_lower:
        return {
            "risk_explanation": (
                f"Given the composite risk score of {risk_score}/100 and threat classification of {threat_type}, "
                f"the AI investigation agent recommends {recommended_action}. "
                f"{'This transaction should be held pending additional verification.' if recommended_action == 'HOLD' else 'This transaction should be reviewed by a risk analyst.' if recommended_action == 'REVIEW' else 'This transaction appears consistent with normal behavior and can be approved.'} "
                f"This is a recommendation — human approval is required for any final action."
            ),
            "attack_story": attack_story,
            "threat_type": threat_type,
            "confidence": 0.93,
            "recommended_action": recommended_action,
            "investigation_summary": f"Recommendation: {recommended_action}. Rationale: {risk_score}/100 score, {risk_level} risk level. Human approval required.",
            "demo": True,
        }

    # Generic response
    return _get_fallback_response(analysis, txn)


def _call_external_ai(api_url: str, api_key: str, analysis: dict, txn: dict, question: str) -> dict:
    """
    Call external AI API if configured.

    The AI receives only structured risk-engine signals — never raw user data.
    Returns structured JSON with cautious language.
    """
    import json
    import urllib.request

    system_prompt = (
        "You are a payment risk investigation AI assistant. "
        "You receive structured transaction and risk-engine signals. "
        "You must ONLY reason over the supplied signals — never invent transaction data. "
        "Use cautious language: 'possible', 'potential', 'likely'. "
        "Never claim a transaction is definitely fraudulent. "
        "Return JSON with: risk_explanation, attack_story, threat_type, confidence (0-1), "
        "recommended_action (APPROVE/REVIEW/HOLD), investigation_summary. "
        "All output is advisory — human approval is required for any action."
    )

    user_content = json.dumps({
        "question": question,
        "transaction": {
            "id": txn.get("id"),
            "amount": txn.get("amount"),
            "currency": txn.get("currency", "₹"),
            "location": txn.get("location"),
        },
        "risk_analysis": {
            "risk_score": analysis.get("risk_score"),
            "risk_level": analysis.get("risk_level"),
            "threat_type": analysis.get("threat_type"),
            "recommended_action": analysis.get("recommended_action"),
            "triggered_signals": [s["label"] for s in analysis.get("risk_factors", []) if s.get("triggered")],
            "attack_story": analysis.get("attack_story"),
        },
    })

    payload = json.dumps({
        "model": "gpt-4o-mini",
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_content},
        ],
        "temperature": 0.3,
        "max_tokens": 500,
    }).encode()

    req = urllib.request.Request(
        api_url,
        data=payload,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
        },
        method="POST",
    )

    with urllib.request.urlopen(req, timeout=15) as resp:
        result = json.loads(resp.read().decode())

    content = result["choices"][0]["message"]["content"]
    try:
        parsed = json.loads(content)
    except json.JSONDecodeError:
        return _get_fallback_response(analysis, txn)

    parsed["demo"] = False
    return parsed

File: src/lib/aiService.ts
import type { AIResponse, RiskAnalysis, Transaction } from '@/types';

const DEMO_RESPONSES: Record<string, (analysis: RiskAnalysis, txn: Transaction) => AIResponse> = {
  'Why was this transaction flagged?': (analysis, txn) => ({
    risk_explanation: `This transaction was flagged because ${analysis.risk_factors
      .filter((f) => f.triggered)
      .map((f) => f.label.toLowerCase())
      .join(', ')} were detected. The composite risk score is ${analysis.risk_score}/100, classified as ${analysis.risk_level}. The most significant contributing signal is ${analysis.risk_factors
      .filter((f) => f.triggered)
      .sort((a, b) => b.deviation - a.deviation)[0]?.label || 'behavioral deviation'}.`,
    attack_story: analysis.attack_story,
    threat_type: analysis.threat_type,
    confidence: 0.87,
    recommended_action: analysis.recommended_action,
    investigation_summary: `Transaction ${txn.id} for ${txn.currency}${txn.amount.toLocaleString()} from ${txn.location} triggered ${analysis.risk_factors.filter((f) => f.triggered).length} risk signals. Composite score: ${analysis.risk_score}/100. Threat classification: ${analysis.threat_type}.`,
    demo: true,
  }),

  'What are the strongest risk signals?': (analysis) => ({
    risk_explanation: `The strongest risk signals, ranked by deviation severity, are: ${analysis.risk_factors
      .filter((f) => f.triggered)
      .sort((a, b) => b.deviation - a.deviation)
      .slice(0, 3)
      .map((f, i) => `${i + 1}. ${f.label} (${f.deviation}% deviation)`)
      .join(', ')}. These signals collectively contribute ${(analysis.risk_factors
      .filter((f) => f.triggered)
      .reduce((sum, f) => sum + f.weight, 0))} points to the weighted risk score.`,
    attack_story: analysis.attack_story,
    threat_type: analysis.threat_type,
    confidence: 0.91,
    recommended_action: analysis.recommended_action,
    investigation_summary: `Top 3 signals: ${analysis.risk_factors.filter((f) => f.triggered).sort((a, b) => b.deviation - a.deviation).slice(0, 3).map((f) => f.label).join(', ')}.`,
    demo: true,
  }),

  'Summarize this investigation.': (analysis, txn) => ({
    risk_explanation: `Investigation summary for transaction ${txn.id}: A ${txn.currency}${txn.amount.toLocaleString()} transaction from ${txn.location} on ${new Date(txn.timestamp).toLocaleString()}. The risk engine computed a score of ${analysis.risk_score}/100 (${analysis.risk_level}), identifying ${analysis.risk_factors.filter((f) => f.triggered).length} triggered signals. The AI investigation agent classifies this as a potential ${analysis.threat_type} and recommends ${analysis.recommended_action}.`,
    attack_story: analysis.attack_story,
    threat_type: analysis.threat_type,
    confidence: 0.89,
    recommended_action: analysis.recommended_action,
    investigation_summary: `Transaction ${txn.id} | ${txn.currency}${txn.amount.toLocaleString()} | ${txn.location} | Score: ${analysis.risk_score}/100 | Level: ${analysis.risk_level} | Threat: ${analysis.threat_type} | Action: ${analysis.recommended_action} | Signals: ${analysis.risk_factors.filter((f) => f.triggered).length}`,
    demo: true,
  }),

  'What attack pattern is possible?': (analysis) => ({
    risk_explanation: `Based on the observed signal pattern — ${analysis.risk_factors
      .filter((f) => f.triggered)
      .map((f) => f.label)
      .join(', ')} — the AI agent identifies a possible ${analysis.threat_type}. This pattern is consistent with known attack methodologies where ${analysis.threat_type.toLowerCase().includes('takeover') ? 'an attacker gains access to a legitimate account from a new device and location, attempts authentication, and rapidly executes high-value transactions' : analysis.threat_type.toLowerCase().includes('velocity') ? 'multiple transactions are executed in rapid succession to exploit a compromised account before detection' : 'behavioral anomalies suggest deviation from established patterns'}. This is a probabilistic assessment, not a definitive determination.`,
    attack_story: analysis.attack_story,
    threat_type: analysis.threat_type,
    confidence: 0.84,
    recommended_action: analysis.recommended_action,
    investigation_summary: `Possible attack pattern: ${analysis.threat_type}. Confidence: 84%. Signals: ${analysis.risk_factors.filter((f) => f.triggered).map((f) => f.key).join(', ')}.`,
    demo: true,
  }),

  'What action do you recommend?': (analysis) => ({
    risk_explanation: `Given the composite risk score of ${analysis.risk_score}/100 and threat classification of ${analysis.threat_type}, the AI investigation agent recommends ${analysis.recommended_action}. ${analysis.recommended_action === 'HOLD' ? 'This transaction should be held pending additional verification (e.g., step-up authentication, manual review by a risk analyst). The combination of triggered signals suggests elevated fraud risk.' : analysis.recommended_action === 'REVIEW' ? 'This transaction should be flagged for manual review by a risk analyst before processing. Some behavioral anomalies were detected but do not meet the threshold for an automatic hold.' : 'This transaction appears consistent with normal behavior and can be approved with standard monitoring.'} This is a recommendation — human approval is required for any final action.`,
    attack_story: analysis.attack_story,
    threat_type: analysis.threat_type,
    confidence: 0.93,
    recommended_action: analysis.recommended_action,
    investigation_summary: `Recommendation: ${analysis.recommended_action}. Rationale: ${analysis.risk_score}/100 score, ${analysis.risk_level} risk level. Human approval required.`,
    demo: true,
  }),
};

const SUGGESTED_QUESTIONS = [
  'Why was this transaction flagged?',
  'What are the strongest risk signals?',
  'Summarize this investigation.',
  'What attack pattern is possible?',
  'What action do you recommend?',
];

export function getSuggestedQuestions(): string[] {
  return SUGGESTED_QUESTIONS;
}

export async function askSentinel(
  question: string,
  analysis: RiskAnalysis,
  txn: Transaction
): Promise<AIResponse> {
  // Demo mode — no external API call
  // In production this would call the backend edge function
  await new Promise((r) => setTimeout(r, 600 + Math.random() * 400));

  const handler = DEMO_RESPONSES[question];
  if (handler) {
    return handler(analysis, txn);
  }

  // Generic response for custom questions
  return {
    risk_explanation: `Based on the available risk signals for transaction ${txn.id}, the risk engine computed a score of ${analysis.risk_score}/100. ${analysis.risk_factors.filter((f) => f.triggered).length} signals were triggered: ${analysis.risk_factors.filter((f) => f.triggered).map((f) => f.label).join(', ')}. The threat is classified as ${analysis.threat_type} with a recommended action of ${analysis.recommended_action}.`,
    attack_story: analysis.attack_story,
    threat_type: analysis.threat_type,
    confidence: 0.85,
    recommended_action: analysis.recommended_action,
    investigation_summary: `Transaction ${txn.id} | Score: ${analysis.risk_score}/100 | ${analysis.threat_type} | ${analysis.recommended_action}`,
    demo: true,
  };
}

export function getFallbackAIResponse(analysis: RiskAnalysis, txn: Transaction): AIResponse {
  return {
    risk_explanation: `AI service is currently unavailable. Based on the rule-based risk engine, transaction ${txn.id} has a score of ${analysis.risk_score}/100 (${analysis.risk_level}). ${analysis.risk_factors.filter((f) => f.triggered).length} signals triggered. Manual review is recommended.`,
    attack_story: analysis.attack_story,
    threat_type: analysis.threat_type,
    confidence: 0.5,
    recommended_action: analysis.recommended_action,
    investigation_summary: `Fallback analysis | ${txn.id} | ${analysis.risk_score}/100 | ${analysis.threat_type} | ${analysis.recommended_action}`,
    demo: true,
  };
}
